reps/2022-04-07-serialization.md

```diff
@@ -0,0 +1,314 @@
+
+    def deserialize(self, in_band_buffer: bytes,
+                    oob_iterator: Iterator[memoryview]) -> bytearray:
+        return next(oob_iterator).obj
+```
+
+Now `bytearray` objects will be out-of-band serialized.
+
+In the same process, no copy will happen to out-of-band buffers.
+
+For example, in the following code, no copy will happen:
+
+```python
+b = bytearray(b'\x01\x02\x03')
+serialization_result: RaySerializationResult = ray.serialize(b)
+deserialized_bytes: bytearray = ray.deserialize(serialization_result)
+```
+
+`deserialized_bytes` and `b` will be backed by the same memory.
+
+When the serialized result needs to go through network, for example, in cross-language situation, only 1 copy will happen. That is from `RaySerializationResult` to the network buffer.
+
+### Nested Serialization
+
+Let's take `List[bytearray]` as an example to see how to handle nested types. We've already implemented a serializer for `bytearray` above.
+
+Note that in real world we might simply wrap Msgpack to do serialization for common classes. Here we rebuild the wheel only to show how it works.
+
+```python
+class ListSerializer(RaySerializer):
+
+    def serialize(self, list_object: List[bytearray]) -> RaySerializationResult:
+        result = RaySerializationResult()
+        result.in_band_buffer = bytearray()
+        # Total element count.
+        result.in_band_buffer.extend(to_bytes(len(list_object))))
+        for element in list_object:
+            element_res = ray.serialize(element)
+            # in band buffer length
+            result.in_band_buffer.extend(to_bytes(len(element_res.in_band_buffer)))
+            # in band buffer
+            result.in_band_buffer.extend(element_res.in_band_buffer)
+            # append out of band buffer
+            result.out_of_band_buffers += element_res.out_of_band_buffers
+        return result
+
+    def deserialize(self, in_band_buffer: bytes,
+                    oob_iterator: Iterator[memoryview]) -> List[bytearray]:
+        result = []
+        pos = 0
+        # read element count
+        count, pos = get_int(in_band_buffer, pos)
+        for i in range(count):
+            # read element size
+            element_size, pos = get_int(in_band_buffer, pos)
+            # read element in-band data
+            element_in_band_data = in_band_buffer[pos: pos + element_size]
+            pos = pos + element_size
+            # nested serialization here
+            element = ray.deserialize(element_in_band_data, oob_iterator)
+            result.append(element)
+        return result
+```
+
+### Fallback Serializer
+
+By default, if no serializer is registered, use serializer in current Ray's version as the fallback. For example, pickle5 in Python, FST in Java.
+
+We'll register a fallback serializer to Ray. For example, pickle5 fallback serializer will be something like the following, all OOB buffers of pickle5 will be wrapped to Ray's OOB buffer.
```

Comment:
What does it exactly mean by fallback here? Does it mean it's used by unregistered classes or is it used when failure happens from the original serializer?

Reply:
it's used by unregistered classes. I changed it to "language default serializer" for clarity.